File: xiaosa/py/芒果视频.py
# -*- coding: utf-8 -*-
# by @嗷呜
import sys
import time
from concurrent.futures import ThreadPoolExecutor, as_completed
sys.path.append('..')
from base.spider import Spider

class Spider(Spider):

    def init(self, extend=""):
        pass

    def getName(self):
        pass

    def isVideoFormat(self, url):
        pass

    def manualVideoCheck(self):
        pass

    def destroy(self):
        pass

    rhost='https://www.mgtv.com'

    host='https://pianku.api.mgtv.com'

    vhost='https://pcweb.api.mgtv.com'

    mhost='https://dc.bz.mgtv.com'

    shost='https://mobileso.bz.mgtv.com'

    headers = {
        'User-Agent': 'Mozilla/5.0 (Windows NT 10.0; Win64; x64; ) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/126.0.6478.61 Chrome/126.0.6478.61 Not/A)Brand/8  Safari/537.36',
        'origin': rhost,
        'referer': f'{rhost}/'
    }

    def homeContent(self, filter):
        result = {}
        cateManual = {
            "电影": "3",
            "电视剧": "2",
            "综艺": "1",
            "动画": "50",
            "少儿": "10",
            "纪录片": "51",
            "教育": "115"
        }
        classes = []
        filters = {}
        for k in cateManual:
            classes.append({
                'type_name': k,
                'type_id': cateManual[k]
            })
        with ThreadPoolExecutor(max_workers=len(classes)) as executor:
            results = executor.map(self.getf, classes)
            for id, ft in results:
                if len(ft):filters[id] = ft
        result['class'] = classes
        result['filters'] = filters
        return result

    def homeVideoContent(self):
        data=self.fetch(f'{self.mhost}/dynamic/v1/channel/index/0/0/0/1000000/0/0/17/1354?type=17&version=5.0&t={str(int(time.time()*1000))}&_support=10000000', headers=self.headers).json()
        videoList = []
        for i in data['data']:
            if i.get('DSLList') and len(i['DSLList']):
                for j in i['DSLList']:
                    if j.get('data') and j['data'].get('items') and len(j['data']['items']):
                        for k in j['data']['items']:
                            videoList.append({
                                'vod_id': k["videoId"],
                                'vod_name': k['videoName'],
                                'vod_pic': k['img'],
                                'vod_year': k.get('cornerTitle'),
                                'vod_remarks': k.get('time') or k.get('desc'),
                            })
        return {'list':videoList}

    def categoryContent(self, tid, pg, filter, extend):
        body={
            'allowedRC': '1',
            'platform': 'pcweb',
            'channelId': tid,
            'pn': pg,
            'pc': '80',
            'hudong': '1',
            '_support': '10000000'
        }
        body.update(extend)
        data=self.fetch(f'{self.host}/rider/list/pcweb/v3', params=body, headers=self.headers).json()
        videoList = []
        for i in data['data']['hitDocs']:
            videoList.append({
                'vod_id': i["playPartId"],
                'vod_name': i['title'],
                'vod_pic': i['img'],
                'vod_year': (i.get('rightCorner',{}) or {}).get('text') or i.get('year'),
                'vod_remarks': i['updateInfo']
            })
        result = {}
        result['list'] = videoList
        result['page'] = pg
        result['pagecount'] = 9999
        result['limit'] = 90
        result['total'] = 999999
        return result

    def detailContent(self, ids):
        vbody={'allowedRC': '1', 'vid': ids[0], 'type': 'b', '_support': '10000000'}
        vdata=self.fetch(f'{self.vhost}/video/info', params=vbody, headers=self.headers).json()
        d=vdata['data']['info']['detail']
        vod = {
            'vod_name': vdata['data']['info']['title'],
            'type_name': d.get('kind'),
            'vod_year': d.get('releaseTime'),
            'vod_area': d.get('area'),
            'vod_lang': d.get('language'),
            'vod_remarks': d.get('updateInfo'),
            'vod_actor': d.get('leader'),
            'vod_director': d.get('director'),
            'vod_content': d.get('story'),
            'vod_play_from': '芒果TV',
            'vod_play_url': ''
        }
        data,pdata=self.fetch_page_data('1', ids[0],True)
        pagecount=data['data'].get('total_page') or 1
        if int(pagecount)>1:
            pages = list(range(2, pagecount+1))
            page_results = {}
            with ThreadPoolExecutor(max_workers=10) as executor:
                future_to_page = {
                    executor.submit(self.fetch_page_data, page, ids[0]): page
                    for page in pages
                }
                for future in as_completed(future_to_page):
                    page = future_to_page[future]
                    try:
                        result = future.result()
                        page_results[page] = result
                    except Exception as e:
                        print(f"Error fetching page {page}: {e}")
            for page in sorted(page_results.keys()):
                pdata.extend(page_results[page])
        vod['vod_play_url'] = '#'.join(pdata)
        return {'list':[vod]}

    def searchContent(self, key, quick, pg="1"):
        data=self.fetch(f'{self.shost}/applet/search/v1?channelCode=mobile-wxap&q={key}&pn={pg}&pc=10&_support=10000000', headers=self.headers).json()
        videoList = []
        for i in data['data']['contents']:
            if i.get('data') and len(i['data']):
                k = i['data'][0]
                if k.get('vid') and k.get('img'):
                    try:
                        videoList.append({
                            'vod_id': k['vid'],
                            'vod_name': k['title'],
                            'vod_pic': k['img'],
                            'vod_year': (i.get('rightTopCorner',{}) or {}).get('text') or i.get('year'),
                            'vod_remarks': '/'.join(i.get('desc',[])),
                        })
                    except:
                        print(k)
        return {'list':videoList,'page':pg}

    def playerContent(self, flag, id, vipFlags):
        id=f'{self.rhost}{id}'
        return  {'jx':1,'parse': 1, 'url': id, 'header': ''}

    def localProxy(self, param):
        pass

    def getf(self, body):
        params = {
            'allowedRC': '1',
            'channelId': body['type_id'],
            'platform': 'pcweb',
            '_support': '10000000',
        }
        data = self.fetch(f'{self.host}/rider/config/channel/v1', params=params, headers=self.headers).json()
        ft = []
        for i in data['data']['listItems']:
            try:
                value_array = [{"n": value['tagName'], "v": value['tagId']} for value in i['items'] if
                               value.get('tagName')]
                ft.append({"key": i['eName'], "name": i['typeName'], "value": value_array})
            except:
                print(i)
        return body['type_id'], ft

    def fetch_page_data(self, page, id, b=False):
        body = {'version': '5.5.35', 'video_id': id, 'page': page, 'size': '30',
                'platform': '4', 'src': 'mgtv', 'allowedRC': '1', '_support': '10000000'}
        data = self.fetch(f'{self.vhost}/episode/list', params=body, headers=self.headers).json()
        ldata = [f'{i["t3"]}${i["url"]}' for i in data['data']['list']]
        if b:
            return data, ldata
        else:
            return ldata

File: xiaosa/js/芒果视频.js
//小心儿悠悠//
var rule = {
    title: '百忙无果[官]',
    host: 'https://pianku.api.mgtv.com',
    homeUrl: '',
    searchUrl: 'https://mobileso.bz.mgtv.com/msite/search/v2?q=**&pn=fypage&pc=10',
    detailUrl: 'https://pcweb.api.mgtv.com/episode/list?page=1&size=50&video_id=fyid',
    searchable: 2,
    quickSearch: 0,
    filterable: 1,
    multi: 1,
    url: '/rider/list/pcweb/v3?platform=pcweb&channelId=fyclass&pn=fypage&pc=80&hudong=1&_support=10000000&kind=a1&area=a1',
    filter_url: 'year={{fl.year or "all"}}&sort={{fl.sort or "all"}}&chargeInfo={{fl.chargeInfo or "all"}}',
    headers: {
        'User-Agent': 'PC_UA'
    },
    timeout: 5000,
    class_name: '电视剧&电影&综艺&动漫&纪录片&教育&少儿',
    class_url: '2&3&1&50&51&115&10',
    filter: {
        "1": getCommonFilter(),
        "2": getCommonFilter(),
        "3": getCommonFilter(),
        "50": getCommonFilter(),
        "51": getCommonFilter(),
        "115": getCommonFilter()
    },
    limit: 20,
    play_parse: true,
    lazy: $js.toString(() => {
        try {
            let api = input.split("?")[0];
            let response = fetch(api, {
                method: 'get',
                headers: {
                    'User-Agent': 'okhttp/3.14.9',
                    'Content-Type': 'application/x-www-form-urlencoded'
                }
            });
            let bata = JSON.parse(response);
            input = {
                parse: 0,
                url: bata.url.includes("mgtv") ? bata.url : input.split("?")[0],
                jx: bata.url.includes("mgtv") ? 0 : 1,
                danmaku: "http://127.0.0.1:9978/proxy?do=danmu&site=js&url=" + input.split("?")[0]
            };
        } catch {
            input = {
                parse: 0,
                url: input.split("?")[0],
                jx: 1,
                danmaku: "http://127.0.0.1:9978/proxy?do=danmu&site=js&url=" + input.split("?")[0]
            };
        }
    }),
    一级: 'json:data.hitDocs;title;img;updateInfo||rightCorner.text;playPartId',
    二级: $js.toString(() => {
        fetch_params.headers.Referer = "https://www.mgtv.com";
        fetch_params.headers["User-Agent"] = MOBILE_UA;

        let videoId = input.split('video_id=')[1].split('&')[0];
        let infoUrl = `https://pcweb.api.mgtv.com/video/info?allowedRC=1&vid=${videoId}&type=b&_support=10000000`;
        let infoData = JSON.parse(request(infoUrl));

        if (infoData && infoData.data && infoData.data.info) {
            let detail = infoData.data.info.detail || {};
            VOD = {
                vod_name: infoData.data.info.title || "",
                type_name: detail.kind || "",
                vod_year: detail.releaseTime || "",
                vod_area: detail.area || "",
                vod_actor: detail.leader || "",
                vod_director: detail.director || "",
                vod_content: detail.story || "",
                vod_remarks: detail.updateInfo || ""
            };
            if (detail.img) VOD.vod_pic = detail.img;
        }

        let d = [];
        let html = request(input);
        let json = JSON.parse(html);
        let host = "https://www.mgtv.com";
        let ourl = json.data.list.length > 0 ? json.data.list[0].url : json.data.series[0].url;
        if (!/^http/.test(ourl)) ourl = host + ourl;

        fetch_params.headers["User-Agent"] = MOBILE_UA;
        html = request(ourl);
        if (html.includes("window.location =")) {
            ourl = pdfh(html, "meta[http-equiv=refresh]&&content").split("url=")[1];
            html = request(ourl);
        }

        try {
            let details = pdfh(html, ".m-details&&Html").replace(/h1>/, "h6>").replace(/div/g, "br");
            let actor = "",
                director = "",
                time = "";
            if (/播出时间/.test(details)) {
                actor = pdfh(html, "p:eq(5)&&Text").substr(0, 25);
                director = pdfh(html, "p:eq(4)&&Text");
                time = pdfh(html, "p:eq(3)&&Text");
            } else {
                actor = pdfh(html, "p:eq(4)&&Text").substr(0, 25);
                director = pdfh(html, "p:eq(3)&&Text");
                time = "已完结";
            }
            let _img = pd(html, ".video-img&&img&&src");
            let JJ = pdfh(html, ".desc&&Text").split("牛马简介：")[1];
            VOD.vod_name = VOD.vod_name || pdfh(html, ".vt-txt&&Text");
            VOD.type_name = VOD.type_name || pdfh(html, "p:eq(0)&&Text").substr(0, 6);
            VOD.vod_area = VOD.vod_area || pdfh(html, "p:eq(1)&&Text");
            VOD.vod_actor = VOD.vod_actor || actor;
            VOD.vod_director = VOD.vod_director || director;
            VOD.vod_remarks = VOD.vod_remarks || time;
            VOD.vod_pic = VOD.vod_pic || _img;
            VOD.vod_content = VOD.vod_content || JJ;
            if (!VOD.vod_name) VOD.vod_name = VOD.type_name;
        } catch (e) {
            log("获取影片信息发生错误:" + e.message);
        }

        function getRjpg(imgUrl, xs) {
            xs = xs || 3;
            let picSize = /jpg_/.test(imgUrl) ? imgUrl.split("jpg_")[1].split(".")[0] : false;
            let rjpg = false;
            if (picSize) {
                let a = parseInt(picSize.split("x")[0]) * xs;
                let b = parseInt(picSize.split("x")[1]) * xs;
                rjpg = a + "x" + b + ".jpg";
            }
            return /jpg_/.test(imgUrl) && rjpg ? imgUrl.replace(imgUrl.split("jpg_")[1], rjpg) : imgUrl;
        }

        if (json.data.total === 1 && json.data.list.length === 1) {
            let data = json.data.list[0];
            d.push({
                title: data.t4,
                desc: data.t2,
                pic_url: getRjpg(data.img),
                url: "https://www.mgtv.com" + data.url
            });
        } else if (json.data.list.length > 1) {
            for (let i = 1; i <= json.data.total_page; i++) {
                if (i > 1) json = JSON.parse(fetch(input.replace("page=1", "page=" + i), {}));
                json.data.list.forEach(function(data) {
                    if (data.isIntact == "1") {
                        d.push({
                            title: data.t4,
                            desc: data.t2,
                            pic_url: getRjpg(data.img),
                            url: "https://www.mgtv.com" + data.url
                        });
                    }
                });
            }
        } else {
            print(input + "暂无片源");
        }
        VOD.vod_play_from = "芒果TV";
        VOD.vod_play_url = d.map(function(it) {
            return it.title + "$" + it.url;
        }).join("#");
        setResult(d);
    }),
    搜索: $js.toString(() => {
        fetch_params.headers.Referer = "https://www.mgtv.com";
        fetch_params.headers["User-Agent"] = MOBILE_UA;
        let d = [];
        let html = request(input);
        let json = JSON.parse(html);
        json.data.contents.forEach(function(data) {
            if (data.type && data.type == 'media') {
                let item = data.data[0];
                if (item.source === "imgo") {
                    let fyclass = '';
                    try {
                        fyclass = item.rpt.match(/idx=(.*?)&/)[1] + '$';
                    } catch (e) {
                        log(e.message);
                    }
                    d.push({
                        title: item.title.replace(/<B>|<\/B>/g, ''),
                        img: item.img || '',
                        content: '',
                        desc: item.desc.join(','),
                        url: fyclass + item.url.match(/.*\/(.*?)\.html/)[1]
                    });
                }
            }
        });
        setResult(d);
    })
};

function getCommonFilter() {
    return [{
        "key": "chargeInfo",
        "name": "付费类型",
        "value": [{
                "n": "全部",
                "v": "all"
            },
            {
                "n": "免费",
                "v": "b1"
            },
            {
                "n": "vip",
                "v": "b2"
            },
            {
                "n": "VIP用券",
                "v": "b3"
            },
            {
                "n": "付费点播",
                "v": "b4"
            }
        ]
    }, {
        "key": "sort",
        "name": "排序",
        "value": [{
                "n": "最新",
                "v": "c1"
            },
            {
                "n": "最热",
                "v": "c2"
            },
            {
                "n": "知乎高分",
                "v": "c4"
            }
        ]
    }, {
        "key": "year",
        "name": "年代",
        "value": [{
                "n": "全部",
                "v": "all"
            },
            {
                "n": "2025",
                "v": "2025"
            },
            {
                "n": "2024",
                "v": "2024"
            },
            {
                "n": "2023",
                "v": "2023"
            },
            {
                "n": "2022",
                "v": "2022"
            },
            {
                "n": "2021",
                "v": "2021"
            },
            {
                "n": "2020",
                "v": "2020"
            },
            {
                "n": "2019",
                "v": "2019"
            },
            {
                "n": "2018",
                "v": "2018"
            },
            {
                "n": "2017",
                "v": "2017"
            },
            {
                "n": "2016",
                "v": "2016"
            },
            {
                "n": "2015",
                "v": "2015"
            },
            {
                "n": "2014",
                "v": "2014"
            },
            {
                "n": "2013",
                "v": "2013"
            },
            {
                "n": "2012",
                "v": "2012"
            },
            {
                "n": "2011",
                "v": "2011"
            },
            {
                "n": "2010",
                "v": "2010"
            },
            {
                "n": "2009",
                "v": "2009"
            },
            {
                "n": "2008",
                "v": "2008"
            },
            {
                "n": "2007",
                "v": "2007"
            },
            {
                "n": "2006",
                "v": "2006"
            },
            {
                "n": "2005",
                "v": "2005"
            },
            {
                "n": "2004",
                "v": "2004"
            }
        ]
    }];
}